callisto/delivery/models.py

```diff
@@ -157,7 +168,31 @@ def decrypted_report(self, key):
         Raises:
           CryptoError: If the key and saved salt fail to decrypt the record.
         """
-        return _decrypt_report(salt=self.salt, key=key, encrypted=self.encrypted)
+        iterations = None
+        harden_runtime = False
+        hasher = identify_hasher(self.encode_prefix)
+
+        if self.encode_prefix == '':
+            # this will only be used in the case of entries made before encode prefixes were used
+            iterations = ORIGINAL_KEY_ITERATIONS
+            salt = self.salt
+            if hasher.iterations > iterations:
+                harden_runtime = True
+        else:
+            salt = self.encode_prefix.rsplit('$', 1)[1]
+
+        # if pbkdf2 is used and the number of iterations changes
+        if self.encode_prefix != '' and hasher.algorithm == 'pbkdf2_sha256' and hasher.must_update(self.encode_prefix):
+            iterations = self.encode_prefix.split('$')[1]
+            harden_runtime = True
+
+        encoded = hasher.encode(key, salt, iterations=iterations)
+        if harden_runtime:
+            hasher.harden_runtime(key, encoded)
+
+        prefix, stretched_key = hasher.split_encoded(encoded)
+
+        return _decrypt_report(salt=self.salt, stretched_key=stretched_key, encrypted=self.encrypted)
```

Comment:
I'm thinking both here and in the get_match method we should try to decrypt first and then update the hash, so that we don't go through that process with a bad key